### PR TITLE
Added benchmarks for OSU Scatter, Gather, Broadcast and Allgather

### PR DESCRIPTION
**Pull request by binhudakhalid**
mpiexec -np 2 julia --project GatherTest.jl 
Running benchmark OSU Gather with type Float32 on 2 MPI ranks

size (bytes),iteratons,min_time (seconds),max_time (seconds),avg_time (seconds)
0,262144,1.9046819943469018e-7,4.0790109778754413e-7,2.9918464861111715e-7
4,262144,5.199644874664955e-7,5.25589712196961e-7,5.227770998317283e-7
8,262144,4.6711011236766353e-7,4.7371031541842967e-7,4.704102138930466e-7
16,262144,4.947023626300506e-7,4.998610165785067e-7,4.972816896042787e-7
32,262144,5.29346834809985e-7,5.343299562809989e-7,5.31838395545492e-7
64,262144,5.943220457993448e-7,6.01074134465307e-7,5.976980901323259e-7

---------------------------------------------------------------------
mpiexec -np 2 julia --project BroadcastTest.jl 
Running benchmark OSU Broadcast with type Float32 on 2 MPI ranks

size (bytes),iteratons,min_time (seconds),max_time (seconds),avg_time (seconds)
0,262144,6.54845280223526e-8,6.597019819309935e-8,6.572736310772598e-8
4,262144,2.744709490798414e-7,2.749602572293952e-7,2.747156031546183e-7
8,262144,2.455853973515332e-7,2.469450919306837e-7,2.4626524464110844e-7
16,262144,2.4846940505085513e-7,2.494007276254706e-7,2.4893506633816287e-7
32,262144,2.720234988373704e-7,2.7476016839500517e-7,2.733918336161878e-7
64,262144,2.4856308300513774e-7,2.5058852770598605e-7,2.495758053555619e-7

----------------------------------------------------------------------
mpiexec -np 2 julia --project ScatterTest.jl 
Running benchmark OSU Scatter with type Float32 on 2 MPI ranks

size (bytes),iteratons,min_time (seconds),max_time (seconds),avg_time (seconds)
0,262144,1.4374836609931663e-7,3.7523568607866764e-7,2.5949202608899213e-7
4,262144,4.1648399928817526e-7,4.292287485441193e-7,4.228563739161473e-7
8,262144,3.492505129543133e-7,3.581908458727412e-7,3.5372067941352725e-7
16,262144,4.5165234041633084e-7,4.6038348955335096e-7,4.560179149848409e-7
32,262144,3.513405317789875e-7,3.7183599488344043e-7,3.6158826333121397e-7
64,262144,5.335659807315096e-7,5.443971531349234e-7,5.389815669332165e-7

----------------------------------------------------------------------
mpiexec -np 2 julia --project GatherTest.jl 
Running benchmark OSU Gather with type Float32 on 2 MPI ranks

size (bytes),iteratons,min_time (seconds),max_time (seconds),avg_time (seconds)
0,262144,5.870051609235816e-7,5.875353963347152e-7,5.872702786291484e-7
4,262144,7.041053322609514e-7,7.045364327495918e-7,7.043208825052716e-7
8,262144,6.546215445268899e-7,6.560039764735848e-7,6.553127605002373e-7
16,262144,7.176595318014733e-7,7.182870831456967e-7,7.17973307473585e-7
32,262144,7.493954399251379e-7,7.49487298890017e-7,7.494413694075774e-7
64,262144,6.680411388515495e-7,6.72624992148485e-7,6.703330655000173e-7

### CODE REVIEW

File: src/osu_allgather.jl [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        MPI.Allgather!(send_buffer, UBuffer(recv_buffer, Cint(bufsize), Cint(nranks), MPI.Datatype(T)), comm)
+        toc = MPI.Wtime()
+        timer += toc - tic
+    end
+    MPI.Barrier(comm)
+    avgtime = timer / iters
+    return avgtime
+end
+
+benchmark(bench::OSUAllgather) = run_osu_collective(bench, osu_allgather, bench.conf)
```
Comment (giordano) [inline src/osu_allgather.jl:38]: Missing newline at the end of the file:
```suggestion
benchmark(bench::OSUAllgather) = run_osu_collective(bench, osu_allgather, bench.conf)

```

File: src/osu_broadcast.jl [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        MPI.Bcast!(send_buffer, 0 , comm)
+        toc = MPI.Wtime()
+        timer += toc - tic
+    end
+    MPI.Barrier(comm)
+    avgtime = timer / iters
+    return avgtime
+end
+
+benchmark(bench::OSUBroadcast) = run_osu_collective(bench, osu_broadcast, bench.conf)
```
Comment (giordano) [inline src/osu_broadcast.jl:33]: Missing newline at the end of the file:
```suggestion
benchmark(bench::OSUBroadcast) = run_osu_collective(bench, osu_broadcast, bench.conf)

```

File: src/osu_gather.jl [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        end
+        toc = MPI.Wtime()
+        timer += toc - tic
+    end
+    MPI.Barrier(comm)
+    avgtime = timer / iters
+    return avgtime
+end
+
+benchmark(bench::OSUGather) = run_osu_collective(bench, osu_gather, bench.conf)
```
Comment (giordano) [inline src/osu_gather.jl:42]: Missing newline at the end of the file:
```suggestion
benchmark(bench::OSUGather) = run_osu_collective(bench, osu_gather, bench.conf)

```

File: src/osu_scatter.jl [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        end
+        toc = MPI.Wtime()
+        timer += toc - tic
+    end
+    MPI.Barrier(comm)
+    avgtime = timer / iters
+    return avgtime
+end
+
+benchmark(bench::OSUScatter) = run_osu_collective(bench, osu_scatter, bench.conf)
```
Comment (giordano) [inline src/osu_scatter.jl:42]: Missing newline at the end of the file:
```suggestion
benchmark(bench::OSUScatter) = run_osu_collective(bench, osu_scatter, bench.conf)

```

File: test/runtests.jl [hunk 5]
```diff
@@ -1,87 +1,93 @@
-using Test
```
Comment (giordano) [inline test/runtests.jl:1]: You appear to have changed the entire file, this isn't good.  Did you change the line-ending of these files?

File: src/osu_allgather.jl [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        "OSU Gather",
+    )
+end
+
+function osu_allgather(T::Type, bufsize::Int, iters::Int, comm::MPI.Comm)
+    rank = MPI.Comm_rank(comm)
+    nranks = MPI.Comm_size(comm)
+    send_buffer = ones(T, bufsize)
+    recv_buffer = zeros(T, bufsize * nranks)
+    root = 0
```
Comment (giordano) [inline src/osu_allgather.jl:23]: This is unused?

File: src/osu_allgather.jl [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    timer = 0.0
+
+    MPI.Barrier(comm)
+    for _ in 1:iters
+        tic = MPI.Wtime()
+        MPI.Allgather!(send_buffer, UBuffer(recv_buffer, Cint(bufsize), Cint(nranks), MPI.Datatype(T)), comm)
```
Comment (giordano) [inline src/osu_allgather.jl:29]: I believe you can also do
```suggestion
        MPI.Allgather!(send_buffer, recv_buffer, comm)
```

File: src/osu_gather.jl [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    send_buffer = ones(T, bufsize)
+    recv_buffer = zeros(T, bufsize * nranks)
+    root = 0
+    timer = 0.0
+
+    MPI.Barrier(comm)
+    for _ in 1:iters
+        tic = MPI.Wtime()
+        if rank == root   
+            MPI.Gather!(MPI.IN_PLACE, UBuffer(recv_buffer, Cint(bufsize) , nranks , MPI.Datatype(T)), comm; root)
```
Comment (giordano) [inline src/osu_gather.jl:30]: Also this one can be simplified to
```suggestion
            MPI.Gather!(MPI.IN_PLACE, recv_buffer, comm; root)
```

File: src/osu_scatter.jl [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    send_buffer = ones(T, bufsize * nranks)
+    recv_buffer = zeros(T, bufsize)
+    root = 0
+    timer = 0.0
+
+    MPI.Barrier(comm)
+    for _ in 1:iters
+        tic = MPI.Wtime()
+        if rank == root   
+            MPI.Scatter!(UBuffer(send_buffer, Cint(bufsize), nranks,MPI.Datatype(T)), recv_buffer, comm; root=root)
```
Comment (giordano) [inline src/osu_scatter.jl:30]: More simplification
```suggestion
            MPI.Scatter!(send_buffer, recv_buffer, comm; root)
```

File: src/osu_scatter.jl [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+        else
+            MPI.Scatter!(nothing, recv_buffer, comm; root=root)
```
Comment (giordano) [inline src/osu_scatter.jl:32]: For consistency with the style of the rest of the code
```suggestion
            MPI.Scatter!(nothing, recv_buffer, comm; root)
```

**Review by giordano — COMMENTED** [review]
Thanks a lot!

File: src/osu_gather.jl (hunk 8 above)
Comment (binhudakhalid) [reply src/osu_gather.jl:30]: When I changed the parameters of the function. It complains that " no method matching Gather!(::MPI.InPlace, ::Vector{Float32}, ::Int64, ::MPI.Comm)".

`ERROR: LoadError: MethodError: no method matching Gather!(::MPI.InPlace, ::Vector{Float32}, ::Int64, ::MPI.Comm)
Closest candidates are:
  Gather!(::Any, !Matched::MPI.UBuffer, ::Integer, ::MPI.Comm) at ~/.julia/packages/MPI/WsAaN/src/collective.jl:249
  Gather!(!Matched::Union{AbstractArray, Ref}, ::AbstractArray, ::Integer, ::MPI.Comm) at ~/.julia/packages/MPI/WsAaN/src/collective.jl:251
  Gather!(::Any, !Matched::Nothing, ::Integer, ::MPI.Comm) at ~/.julia/packages/MPI/WsAaN/src/collective.jl:253
  ...`

File: src/osu_scatter.jl (hunk 9 above)
Comment (binhudakhalid) [reply src/osu_scatter.jl:30]: After the changes, it generates the error "LoadError: DivideError: integer division error"

File: test/runtests.jl (hunk 5 above)
Comment (binhudakhalid) [reply test/runtests.jl:1]: Sorry, It was due to my Visual Studio Code setting

File: src/osu_allgather.jl (hunk 7 above)
Comment (binhudakhalid) [reply src/osu_allgather.jl:27]: After the modification it give the error "LoadError: ERROR: LoadError: DivideError: integer division error"

File: src/osu_gather.jl (hunk 8 above)
Comment (giordano) [reply src/osu_gather.jl:30]: Right, can use the array if the size can be inferred from `sendbuf`, but in this case it isn't another array.

File: src/osu_allgather.jl (hunk 7 above)
Comment (giordano) [reply src/osu_allgather.jl:27]: Uhm, is it because size 0 isn't handled nicely?  This sounds like a possible bug in `MPI.jl`.

**Comment by giordano** [comment]
Thank you so much!

**Pull request merged**